### PR TITLE
[GuestLoader]: Add autoboot for guestloader

### PR DESCRIPTION
**Pull request by in1004kyu**
1. Add CLI (command.*)
2. Add autoboot (main.c)
3. Fix guestloader's memory map(Add sp for system/user mode)
4. Add writel/readl function in asm_io.h

CLI is worked on RTSM. Arndale is not supproted.

### CODE REVIEW

**Comment by kesl** [comment]
CI : Test PASSed.
Refer to this link for build results: http://keslproject.iptime.org:8082/job/k-hypervisor_pull_request/137/

File: common/guest/command/cmd_menu.c [hunk 1]
```diff
@@ -0,0 +1,124 @@
+#include <cmd_menu.h>
+#include <loadguest.h>
+#include <version.h>
+
+enum cmd_val {
```
Comment (Wonseok) [inline common/guest/command/cmd_menu.c:5]: Please make meaningful name.

File: common/guest/command/cmd_menu.c [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    HELP,
+    BOOT,
+    BANNER,
+    STATUS,
+    BADNUM = -1
```
Comment (Wonseok) [inline common/guest/command/cmd_menu.c:10]: What is the bad num? Here is no relation between BADNUM and others.

File: common/guest/command/cmd_menu.c [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+};
+
+struct cmd {
+    char *key;
+    int val;
+};
+
+static struct cmd cmdtbl[] = {
```
Comment (Wonseok) [inline common/guest/command/cmd_menu.c:18]: Think again. Do you really need to use this structure?

Comment (Wonseok) [inline common/guest/command/cmd_menu.c:18]: The cmdtbl is not good name.

Comment (Wonseok) [inline common/guest/command/cmd_menu.c:18]: Why don't you declare this variable statically?

File: common/guest/command/cmd_menu.c [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    {"banner", BANNER},
+    {"status", STATUS},
+    {"s", STATUS}
+};
+
+#define NUM_KEY (sizeof(cmdtbl) / sizeof(struct cmd))
+#define MAX_ARG_SIZE    32
+#define NULL '\0'
+
+/** @brief Return command to integer type from string type.
```
Comment (Wonseok) [inline common/guest/command/cmd_menu.c:32]: s/Return/Returns
s/integer type/integer
s/string type/string

File: common/guest/command/cmd_menu.c [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+ *  @param key Command that will be converted to integer type.
```
Comment (Wonseok) [inline common/guest/command/cmd_menu.c:33]: Wrong comment.

File: common/guest/command/cmd_menu.c [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+ *  @return Integer type command value.
```
Comment (Wonseok) [inline common/guest/command/cmd_menu.c:34]: Not integer type, return value is enum and it's defined by you.

File: common/guest/command/cmd_menu.c [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+ */
+enum cmd_val cmd_ktov(char *key)
```
Comment (Wonseok) [inline common/guest/command/cmd_menu.c:36]: ktov?

File: common/guest/command/cmd_menu.c [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+{
+    int i;
+    for (i = 0; i < NUM_KEY; i++) {
+        struct cmd *cmd = cmdtbl + i;
```
Comment (Wonseok) [inline common/guest/command/cmd_menu.c:40]: In this case, you don't need to use loop.

File: common/guest/command/cmd_menu.c [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+ */
+void cmd_usage(void)
+{
+    uart_print("help (h)        - List commands and their usage\n\n"
+               "boot (b)        - Boot guestos\n\n"
+               "banner          - Print the K-Hypervisor banner\n\n"
+               "status (s)      - Print the K-Hypervisor status\n\n");
+}
+
+/** @brief Boots guest os.
```
Comment (Wonseok) [inline common/guest/command/cmd_menu.c:57]: s/Boots guest os/Boots up guest OS in hypervisor.

File: common/guest/command/cmd_menu.c [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+ */
+void cmd_run_guest(void)
```
Comment (Wonseok) [inline common/guest/command/cmd_menu.c:59]: Comment and function name are not the same. boot, run, load.

File: common/guest/command/cmd_menu.c [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+/** @brief Return command to integer type from string type.
+ *  @param key Command that will be converted to integer type.
+ *  @return Integer type command value.
+ */
+enum cmd_val cmd_ktov(char *key)
+{
+    int i;
+    for (i = 0; i < NUM_KEY; i++) {
+        struct cmd *cmd = cmdtbl + i;
+        if (strcmp(cmd->key, key) == 0)
```
Comment (Wonseok) [inline common/guest/command/cmd_menu.c:41]: If user input has a trailing space, then it could handle the case?

File: common/guest/command/cmd_menu.c [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+/** @brief Prints K-Hypervisor's banner.
+ */
+void cmd_print_banner(void)
+{
+    uart_print(BANNER_STRING);
+}
+
+/** @brief Prints K-hypervisor's status.
+ *  @todo (InKyu Han)Not implemented yet
```
Comment (Wonseok) [inline common/guest/command/cmd_menu.c:73]: Doxygen cannot support todo sentence like this: todo(InKyu Han) prints hypervisor status.
If you want to add your name in todo, you don't need to use your full name, use inkyu.

File: common/guest/command/cmd_menu.c [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+                argv[argc] = &line[pos];
+                argc++;
+            }
+            cnt++;
+        }
+        pos++;
+    }
+
+    if (argc) {
+        switch (cmd_ktov(argv[0])) {
```
Comment (Wonseok) [inline common/guest/command/cmd_menu.c:109]: You don't need to add prefix 'cmd_' in all functions in this file.

File: common/guest/command/cmd_menu.h [hunk 14]
```diff
@@ -0,0 +1,15 @@
+#ifndef __CMD_MENU_H__
+#define __CMD_MENU_H__
+
+/**
+ * @brief Executes command.
+ *
```
Comment (Wonseok) [inline common/guest/command/cmd_menu.h:6]: Please add the title here. e.g. command: description.

File: common/guest/command/cmd_menu.h [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+ * help (h)        - List commands and their usage
+ * boot (b)        - Boot guestos
+ * banner          - Print The K-Hypervisor banner
+ * status (s)      - Print The K-Hypervisor status
+ *
+ * @param line The string of command.
```
Comment (Wonseok) [inline common/guest/command/cmd_menu.h:12]: s/The string of command/Command string.

File: common/guest/command/cmd_menu.h [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+ */
+void cmd_exec(char *line);
```
Comment (Wonseok) [inline common/guest/command/cmd_menu.h:14]: s/char *line/char *command or
s/char *line/char *cmd/

File: common/guest/core/c_start.c [hunk 17]
```diff
@@ -27,6 +27,7 @@
 #include <log/uart_print.h>
 #include <gic.h>
 #include <test/tests.h>
+#include <test/test_vtimer.h>
```
Comment (Wonseok) [inline common/guest/core/c_start.c:30]: This pull request is autoboot for guestloader only. Why this file is in this pull request?
You might want to send another pull request.

File: common/guest/core/gic.c [hunk 18]
```diff
@@ -4,7 +4,6 @@
 #include <hvmm_trace.h>
 #include <a15_cp15_sysregs.h>
 #include <armv7_p15.h>
-#include <test/test_vtimer.h>
```
Comment (Wonseok) [inline common/guest/core/gic.c:7]: Ditto.

File: common/guest/loader/loadguest.c [hunk 19]
```diff
@@ -35,3 +35,25 @@ void load_bmguest(uint32_t start_addr)
     /* Jump to 0x8000_0000 */
     asm volatile ("mov pc, %0" : : "r" (start_addr) : "memory", "cc");
 }
+
+void load_guest(uint32_t type)
```
Comment (Wonseok) [inline common/guest/loader/loadguest.c:39]: Type is not good naming in this case.

File: common/guest/loader/loadguest.c [hunk 20]
```diff
@@ -35,3 +35,25 @@ void load_bmguest(uint32_t start_addr)
...
+{
+    copy_loader_next_to_guest();
+    copy_guestos_to_start_addr(START_ADDR);
+    switch (type) {
+    case GUEST_TYPE_BM:
+    case GUEST_TYPE_RTOS:
+        load_bmguest(START_ADDR);
+        break;
+    case GUEST_TYPE_LINUX:
+        load_linuxguest(START_ADDR);
```
Comment (Wonseok) [inline common/guest/loader/loadguest.c:49]: load_linuxguest looks very similar with u-boot. Could you please rewrite everything?

File: common/guest/loader/loadguest.c [hunk 21]
```diff
@@ -35,3 +35,25 @@ void load_bmguest(uint32_t start_addr)
     /* Jump to 0x8000_0000 */
     asm volatile ("mov pc, %0" : : "r" (start_addr) : "memory", "cc");
 }
+
+void load_guest(uint32_t type)
+{
+    copy_loader_next_to_guest();
+    copy_guestos_to_start_addr(START_ADDR);
```
Comment (Wonseok) [inline common/guest/loader/loadguest.c:42]: How do you handle case of multiple guest os?

**Comment by Wonseok** [comment]
Please fix it.

File: common/guest/command/cmd_menu.c (hunk 2 above)
Comment (in1004kyu) [inline common/guest/command/cmd_menu.c:10]: I will fix it "BADCMD"

File: common/guest/command/cmd_menu.c (hunk 1 above)
Comment (in1004kyu) [inline common/guest/command/cmd_menu.c:5]: cmd_name

File: common/guest/command/cmd_menu.c (hunk 7 above)
Comment (in1004kyu) [inline common/guest/command/cmd_menu.c:36]: key to value

File: common/guest/loader/loadguest.c (hunk 21 above)
Comment (in1004kyu) [inline common/guest/loader/loadguest.c:42]: guest loader equals to guest os. Guest loader uses this file.

File: common/guest/command/cmd_menu.c (hunk 3 above)
Comment (in1004kyu) [inline common/guest/command/cmd_menu.c:18]: It is static struct.

Comment (in1004kyu) [inline common/guest/command/cmd_menu.c:18]: If new commend is added or removed, programmer should check array's maxnumber always. It is so inconvenient.
